contracts/erc721/UserMintableERC721.sol

```diff
@@ -36,30 +33,100 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Metadata {
     return anchorRegistry_;
   }
 
+  /**
+  * Returns the values associated with a token
+  * @param anchorId uint256 The ID of the document as identified
+  * by the set up anchorRegistry
+  * @param documentRoot bytes32 The root hash of the merkle proof/doc
+  */
+  function getTokenDetails(uint256 _tokenId)
+  external
+  view
+  returns (
+    uint256 anchorId,
+    bytes32 documentRoot
+  )
+  {
+    return (
+    tokenDetails_[_tokenId].anchorId,
+    tokenDetails_[_tokenId].rootHash
+    );
+  }
+
   /**
    * @dev Constructor function
    * @param _name string The name of this token
    * @param _symbol string The shorthand token identifier
    * @param _anchorRegistry address The address of the anchor registry
-   * @param _mandatoryFields array of field names that are being proved
    * using document root and precise-proofs.
    * that is backing this token's mint method.
    */
   function initialize(
     string memory _name,
     string memory _symbol,
-    address _anchorRegistry,
-    string[] memory _mandatoryFields
+    address _anchorRegistry
   )
   public
   initializer
   {
     anchorRegistry_ = _anchorRegistry;
-    mandatoryFields = _mandatoryFields;
     ERC721.initialize();
+    ERC721Enumerable.initialize();
     ERC721Metadata.initialize(_name, _symbol);
   }
 
+  /**
+   * @dev Mints a token after validating the given merkle proof
+   * and comparing it to the anchor registry's stored hash/doc ID.
+   * @param _to address The recipient of the minted token
+   * @param _tokenId uint256 The ID for the minted token
+   * @param _anchorId bytes32 The ID of the document as identified
+   * by the set up anchorRegistry.
+   * @param _merkleRoot bytes32 The root hash of the merkle proof/doc
+   * @param _tokenURI string The metadata uri
+   * @param _values string[] The values of the leafs that are being proved
+   * using precise-proofs
+   * @param _salts bytes32[] The salts for the field that is being proved
+   * Will be concatenated for proof verification as outlined in
+   * precise-proofs library.
+   * @param _proofs bytes32[][] Documents proofs that are needed
+   * for proof verification as outlined in precise-proofs library.
+   */
+  function mintAnchor(
```

Comment:
@manuelpolzhofer what exactly do you mean here?

Reply:
If I understood it correctly, we would have two contracts.

PaymentObligation and ERC721.

The paymentObligation contract calls the mint method on the erc721 contract.
Inside the paymentObligation we are logging the revealed fields and creating events.

We need to prevent to call the erc721 directly and mint an NFT.
only paymentObligation calls mint on erc721 should be allowed

Reply:
@rdinicut what do you think?

Reply:
We do not an additional check because ERC721 and PO because that will be ciclic dependancy. So a ERC721 deploy is not used for all NFT types just for POs in this case. you first deploy the ERC721 where you configure what AnchorRepo and define the name and token name. This contract contains centrifuge specific logic. Then you deploy the PO that has a configured ERC721. Composition is used because we want to update them independently(ex: we add new business logic in the mint method of the PO we do not need to push an update to ERC721)

Reply:
All PO NFTs can use our ERC721 but implement a different business logic. Have diff fields and validations and they can all point to our deployed contract. If you would have a type different nft(ex: one for invoices) we will need to deploy another ERC721(Mainly for the naming part)

Reply:
You can call it the ERC721 but the address of contract is tied to the NFT. In this case the PO contract not the ERC721

Reply:
but every different NFT would use its own instance of an ERC721 otherwise it would make no sense to me.

What do you mean with cyclic dependancy?

Reply:
Well how would you deploy the contracts if they need each others address?

Reply:
You could do a setter but the idea is you should be able to set that only in the constructor/initializer

Reply:
In theory you can reuse the ERC721 for types of NFT but you would have a ERC721 for each NFT. The you can reuse the implementation with ZOS. We have a different ERC721 proxy but the same implementation

Reply:
When you Deploy an PO now you feed it a configured proxy of the ERC721.

Reply:
Let the PO contract create the erc721.
something like:
```javascript
tokenRegistry_ = new UserMintableERC721(...);
```

Maybe a explorer is monitoring the PO events to show the NFT transfers etc.
If I could call the mint method on the erc721 directly without the `msg.sender == poAddress` check, we wouldn't log the event in the PO contract

Reply:
In the ERC721 i only would store the address of the `msg.sender` in the constructor/init which would be the PO address.

If PO create ERC721.

Reply:
I do not think this enforcing is needed. There is no problem if mint an nft on the ERC721. The NFT registry is the PO. That is just spam. The worst thing that can happen is to reserve some tokenIds for nfts. You need to have a minted anchor and provide a proof to mint an nft so that is not that triavial IMO

Reply:
msg.sender is not the contract because you need to use CALLCODE like in the identity contract to do that and the nft needs to be owned by the actual sender.

Reply:
1. LIC NFT and PAY NFT will not use the same UM NFT instance. They will use different proxies that can point to same UMNFT implementation
2. Yes you can because the storage is on the proxy
3. That can be a problem  yes
 
This is how you deploy a NFT with the current implementation. 
```
   const userMintableERC721 = await create(Object.assign({
        contractAlias: 'UserMintableERC721',
        initMethod: 'initialize',
        initArgs: ["Centrifuge Payment Obligations", "CENT_PAY_OB", anchorRepository.address]
    }, options));
    const paymentObligation = await create(Object.assign({
        contractAlias: 'PaymentObligation',
        initMethod: 'initialize',
        initArgs: [userMintableERC721.address]
    }, options));
```

This code creates zos proxies

Reply:
You can not reuse the UM NFT because each NFT has to have constant name and token. That is stored in the UM NFT proxy

Reply:
Sticking with the sample names from @lucasvo - UM (as the base contract), LIC, and PAY.

I think it is problematic if the UM instance's `mint` can be called directly - regardless if there are different instances for LIC and PAY. The mint method might emit events or do other "unexpected things" that could mess up storage (as @manuelpolzhofer pointed out). Especially, if the LIC and PAY are using the same storage as their underlying UM instances. 

The goal should be that only the PAY and LIC's methods are callable. This will keep the system cleaner.

What's really the benefit of this change to composition? Why is this a "must have" to use composition in this case? If we use inheritance we could still upgrade the LIC and PAY contracts, right? It would just mean we might have to upgrade "more"?

Reply:
I think actually the key point is that in the current contract you could call `mint` on `UM` directly and skip certain checks that `PM` would do and then own a `PM` NFT. 

Because the storage is shared now the `PM` would say yes this token actually exists. So the current design is definitely broken.

Irregardless of this, I'm not sure that requiring deploying two contracts is very helpful because the `UM` contract can't be reused and it needs to be upgraded anyway, why wouldn't you just deploy it as one contract (without using zos)?

Reply:
I agree with you, @lucasvo on the problematic points and agree that the design is flawed in the current setup.
However, we want to make sure that the PAY and LIC contracts would still be upgradeable. From what I understand that's the reason to use zos here. To be able to upgrade those two contracts. 

@rdinicut can we not have upgradeability for the PAY NFT even if we use inheritance from the UM contract?

Reply:
In a nutshell, if we use inheritance we can leverage zos upgrade but not the evm packages. 

If you would use inheritance you can update your contracts but you will need to pull the new  version locally, deploy a new implementation that uses our new smart contract and update your implementation. Also it can have a scaling problem if you have multiple NFTS. Instead of updating just the UM NFT implementation you would have to deploy and update all the implementations. (Ex: 100 NFT contracts = 100 deploys + 100 transactions to change the implementation)

With a composition pattern you do not need to deploy your PAY and LIC contracts. You just need to update the implementation of UM NFT in case Centrifuge pushes a new version of the evm package. (Ex: 100 NFT contracts = 100 transactions to change the UM NFT implementation)

Regarding the tokenID collision problem: I think it's the same use case as eth accounts, anchor ids: we can use high entropy random number generation to avoid collisions. Yes you can call the UM NFT but a NFT value is tied to the contract address. So the worst thing you can do is to reserve a certain tokenID and the mint method will fail in case of collision and you need to generate another tokenID

Reply:
Don't we already use high entropy to avoid collision between nodes? So we should already should handle tokenID collisions in the protocol

Reply:
@rdinicut no, the problem is that when you do UM.mintToken("0x01") and then some other contract checks `PAY.ownerOf("0x01") == msg.sender`, that check would pass even if the attacker never minted a PAY NFT.

Reply:
@lucasvo I just discussed it with @rdinicut.
If we use composition the `PAY` would need to implement all the ERC-721 interface methods to be ERC-721 compatible.
We would have some wrapper which would check if the NFT was minted with the `PAY.mint` method. The `PAY.mint` would store additional data, which tokenID's have been minted by `PAY.mint -> UM.mint` flow.
Every ERC-721 method in `PAY` would check if the `tokenID` exists in `PAY` before calling the `UM` implemenation.

An alternative could be to use the fallback function in `PAY` to forward the interface method calls to `UM`.

Reply:
Let's discuss it f2f with @pstehlik and @rdinicut  later

Reply:
@lucasvo At the moment you can not call Pay.ownerOf because it does not exist. You could implement a fallback method but we need to be mindful of methods that should not fallback and be implemented in the PAY contract. ownerOf, like mint and getTokenDetails is one of those methods.

Reply:
So I think we will have an Interface and Class that PO would extend and I would also rename UserMIntableERC721 to CentrifugeERC721 but I do not think it's the scope of this PR

Reply:
@rdinicut but that's extremely confusing, that the contract doesn't actually even implement erc721 but you have this other contract that does. I don't think that's worth saving one transaction.